### PR TITLE
feat: `getLogs` cheatcode

### PR DESCRIPTION
**Pull request by puma314**
## Motivation

Right now there is a `getRecordedLogs()` cheatcode that allows users to call `vm.recordLogs()` and retrieve subsequent emitted logs for post-processing or other use-cases.

However, there is no way for users to retrieve logs that are emitted outside of the `recordLogs()` context.

A lot developers use one-off Typescript and Go scripts to retrieve logs and then conditionally call other Solidity functions with them. This requires significant duplication of existing Solidity logic in other languages. If these scripts could be replaced with Forge scripts that are able to interact with existing Solidity logic, it would simplify off-chain workflows significantly. The ability to retrieve logs in a similar manner to the `getLogs` RPC within a Forge script is the main blocker to using Forge scripts for these sorts of tasks.

## Solution

Create a new cheatcode `getLogs(uint256 fromBlock, uint256 toBlock, address filterAddress, bytes32[] memory topics)` with the same arguments as the `eth_getLogs` RPC.

If there is an active fork, use the `fork_url` to query the `getLogs` RPC and return the result as a `vm.Logs[]` object. If there is no active fork, then throw an error.

### CODE REVIEW

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:158]: Feel free to also use `?` here

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:148]: This is fine actually!

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:146]: We can extract this body into a function—it'll be easier to maintain down the road

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:169]: Maybe instead of panicking (which is fair in this case, as it should be impossible), we can also `eyre::bail` with an explicit error as to why this _should_ be impossible.

File: testdata/cheats/Fork2.t.sol
Comment (Evalir) [inline testdata/cheats/Fork2.t.sol:209]: sweet yes! i'd also include some comments on how to regen this fixture, if needed.

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:180]: cc @mattsse — ig there's no way to avoid spawning a blocking handle here right?

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:211]: nit re: formatting

you might have format on save enabled—we use the nightly toolchain for clippy and rust, so to make the fmt/clippy jobs pass you might need to do `cargo +nightly fmt` and `cargo +nightly clippy --workspace` before you push.

**Review by Evalir — CHANGES_REQUESTED** [review]
thanks for this! it's going in the right direction. let's keep resolving the TODOs left, and:
- making the test data a fixture
- removing the unneeded comments once todos are solved
- extract the cheatcode itself into a function.

also, left a nit re: formatting.

**Comment by mds1** [comment]
This PR basically adds a new `vm.getLogs` cheat which is just a way to call the `eth_getLogs` RPC method natively inside a test/script, without `vm.ffi`, which is great and definitely valuable

The ability to call RPC methods from tests/scripts has been a request for a long time, often to allow configuring anvil into a specified state from a forge script. So before merging a one-off cheat for this method I think it's worth framing this more broadly around "how to call RPC methods from script/tests".  I see three options:

1. Native cheats for each RPC method. Downside here is that this ends up adding a lot of cheats, and you need bespoke cheats for each chain. 
    - With this approach, we may want to rename this to `vm.eth_getLogs` to match the RPC method—we should think about the naming convention otherwise we pollute the cheatcode namespace with lots of RPC cheats that aren't easily identifiable as "aliases for RPC calls", since that's not immediately clear from `vm.getLogs`
2. A single `vm.rpc(string memory name, bytes memory args) returns (bytes memory returnData)` cheat that takes the method name and ABI-encoded args, which are decoded and sent to the RPC. The return data is ABI-encoded and can be decoded in solidity. This scales nicely for arbitrary RPC calls on any chain
3. A hybrid approach where common/standardized methods have dedicated cheats and we have the `vm.rpc` escape hatch for the rest

I lean towards 2 or 3, and also lean towards just matching the full RPC method name for clarity. Interested to hear other thoughts on this

**Comment by puma314** [comment]
> I lean towards 2 or 3, and also lean towards just matching the full RPC method name for clarity. Interested to hear other thoughts on this

I also strongly support matching the full RPC method name, that's a great suggestion. We can also name the return type as `EthGetLogsResponse[]` (or I'm open to other naming suggestions/conventions here since that name is a bit verbose) because the standard Foundry `vm.Logs` only has a subset of values returned by the RPC and I've found all the returned values to be useful at some point or another.

Between approaches 2 and 3, I lean towards 3 because I think there are strong benefits for type-safety for the most common RPC calls, but the `vm.rpc` escape hatch seems nice for all the rest.

**Comment by mds1** [comment]
> I also strongly support matching the full RPC method name, that's a great suggestion. We can also name the return type as `EthGetLogsResponse[]` (or I'm open to other naming suggestions/conventions here since that name is a bit verbose) because the standard Foundry `vm.Logs` only has a subset of values returned by the RPC and I've found all the returned values to be useful at some point or another.
> 
> Between approaches 2 and 3, I lean towards 3 because I think there are strong benefits for type-safety for the most common RPC calls, but the `vm.rpc` escape hatch seems nice for all the rest.

Great, I'm on board with all of this, as long as @mattsse / @Evalir also agree. Then the one nit we'll have to standardize is the naming conventions:
- `vm.eth_getLogs` vs. `vm.ethGetLogs`
- `EthGetLogsResponse[]` vs. `Eth_GetLogs_Response[]` vs. `EthGetLogs_Response[]`, arguably even `EthGetLogs[]` works

My weakly held opinion here is `vm.eth_getLogs()` (to identically match the RPC name and make it clear this is an RPC query), and `EthGetLogs[]` but I can be convinced of other naming styles

**Comment by puma314** [comment]
If we're all on board with the following, can start working on implementing this!
* Cheatcode named: `vm.eth_getLogs()`
* Response: `EthGetLogs[]`
* General RPC cheatcode "escape hatch": `vm.rpc(string memory name, bytes memory args) returns (bytes memory returnData)`

**Comment by Evalir** [comment]
I'm onboard with the naming—while i think `eth_getLogs` definitely breaks the pattern we've had so far, it's good to mirror the actual rpc naming. Let's do it

**Comment by Evalir** [comment]
whoops @puma314 sorry—marked this as RFR even though youre still cleaning up. Feel free to ping me when you want another review round!

File: testdata/cheats/Fork2.t.sol
Comment (puma314) [reply testdata/cheats/Fork2.t.sol]: Included in `README.md` under `testdata/fixtures/Rpc` folder.

**Comment by puma314** [comment]
@Evalir ready for another round of review! Addressed all of your existing comments for `getLogs` and also made `.rpc(...)` work with a simple test case.

**Comment by Evalir** [comment]
Sweet! Do you think you can resolve the conflicts? @puma314

**Comment by puma314** [comment]
Just fixed @Evalir!

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:266]: I think we could put `4` in a constant along with a comment as to why it needs to be les than 4, to be extra clear

File: testdata/cheats/Fork2.t.sol
Comment (Evalir) [inline testdata/cheats/Fork2.t.sol:229]: can we add a test for a more complex call?

File: testdata/cheats/Vm.sol
Comment (Evalir) [inline testdata/cheats/Vm.sol:537]: let's also document the `rpc` function

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:259]: Let's add docs here

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (Evalir) [inline evm/src/executor/inspector/cheatcodes/fork.rs:335]: Also some docs here

**Review by Evalir — CHANGES_REQUESTED** [review]
looking good! just some more asks re docs and tests.

would love your eyes here @mattsse!

**Review by Evalir — APPROVED** [review]

**Review by Evalir — APPROVED** [review]
aight, took this over the line—just needed a smol fix to the test. Thank you!

**Comment by puma314** [comment]
Thank you so much @Evalir!! Sorry the last few weeks got a little busy so I didn't have time to finish this up. Awesome that this is finally merged in, I'm excited to use it :D

**Pull request merged**
